File: number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int l =0;
        unordered_map<char,int>mp;
        int n = s.length();

        int ans = 0;
        for(int r=0; r<n; r++){
            mp[s[r]]++;
            while(mp.size()>=3){
                ans += (n-r);
                mp[s[l]]--;
                if(mp[s[l]]==0){
                    mp.erase(s[l]);
                }
                l++;
            }
        }

        return ans;
    }
};

File: number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        dict = {}
        r = 0
        l = 0
        n = len(s)
        ans = 0
        while r<n:
            dict[s[r]] = dict.get(s[r],0)+1
            while len(dict)>=3:
                ans += n-r
                dict[s[l]] = dict.get(s[l],0)-1
                if(dict.get(s[l])==0):
                    dict.pop(s[l])
                l+=1
            r+=1
        return ans
        